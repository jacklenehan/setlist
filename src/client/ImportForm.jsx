import React from 'react';
import { connect } from 'react-redux';
import { importTracks } from './actions';

import './ImportForm.css'

let ImportForm = ({ dispatch }) => {
  let textarea;

  return (
    <section className='ImportForm'>
      <h3>import</h3>
      <form
        onSubmit={e => {
          e.preventDefault();
          const value = textarea.value;
          if(!value.trim()) {
            return;
          }
          dispatch(importTracks(value));
        }}
      >
        <textarea
          className='uri-list'
          placeholder='Enter URIs here'
          ref={node => {textarea = node}}
        />
        <button type='submit'>Add to tracks</button>
      </form>
    </section>
  )
}

ImportForm = connect()(ImportForm);

export default ImportForm;
