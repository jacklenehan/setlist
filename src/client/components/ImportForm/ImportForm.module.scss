.root {
  grid-column-start: 1;
  grid-column-end: 3;
  padding: 0 0.5rem;

  form {
    width: 100%;
  }

  textarea {
    box-sizing: border-box;
    width: 100%;
    height: 600px;
    display: block;
  }
}
