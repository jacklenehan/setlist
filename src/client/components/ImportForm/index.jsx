import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'

import { trackActions } from 'Client/redux/track'
import { parseIds } from 'Client/helpers'

import styles from './ImportForm.module.scss'

class ImportForm extends React.Component {
  static propTypes = {
    importTracks: PropTypes.func.isRequired,
  }

  constructor(props) {
    super(props)
    this.textarea = React.createRef()
  }

  handleSubmit = (e) => {
    e.preventDefault()

    const { importTracks } = this.props

    const joinedIds = parseIds(this.textarea.current.value).join(',')
    if (joinedIds.length > 0) {
      importTracks(joinedIds)
    }
  }

  render() {
    return (
      <section className={styles.root}>
        <h3>import</h3>
        <form onSubmit={this.handleSubmit}>
          <textarea placeholder="Enter URIs here" ref={this.textarea} />
          <button type="submit">Add to tracks</button>
        </form>
      </section>
    )
  }
}

ImportForm.propTypes = {
  importTracks: PropTypes.func.isRequired,
}

export default connect(
  null,
  {
    importTracks: trackActions.importTracks,
  },
)(ImportForm)
