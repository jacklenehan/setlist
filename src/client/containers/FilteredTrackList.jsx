import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Moment from 'moment';
import 'moment-duration-format';

import TrackList from '../components/TrackList';

const mapStateToProps = (state, { name }) => {
  const listTracks = state.tracksByList[name].tracks.map(track => state.entities.tracks[track]);
  const totalDuration = listTracks.reduce((acc, track) => acc += track.duration_ms, 0);

  return {
    isTooLong: false,
    formattedDuration: '0:00:00',
    tracks: listTracks,
    name
  }
};

const FilteredTrackList = connect(
  mapStateToProps
)(TrackList);

FilteredTrackList.propTypes = {
  name: PropTypes.string.isRequired
}

export default FilteredTrackList;
