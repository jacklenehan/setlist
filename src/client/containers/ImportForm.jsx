import React from 'react';
import { connect } from 'react-redux';
import { importTracks } from '../actions';
import { parseIds } from '../helpers'

export const ImportForm = ({ dispatch }) => {
  let textarea;
  const handleSubmit = e => {
    e.preventDefault();
    const joinedIds = parseIds(textarea.value).join(',');
    if(joinedIds.length > 0) {
      dispatch(importTracks(joinedIds));
    }
  };

  return (
    <section className='ImportForm'>
      <h3>import</h3>
      <form onSubmit={handleSubmit} >
        <textarea
          className='uri-list'
          placeholder='Enter URIs here'
          ref={node => {textarea = node}}
        />
        <button type='submit'>Add to tracks</button>
      </form>
    </section>
  )
}

export default connect()(ImportForm);
