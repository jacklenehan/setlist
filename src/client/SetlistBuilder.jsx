import React from 'react';
import PropTypes from 'prop-types';
import ImportForm from './ImportForm';

import './App.css';

const SetlistBuilder = ({ showImport, toggleImportForm }) => (
  <div className={`App ${showImport ? 'show' : 'hide'}-import-form`}>
    {
      showImport
      ? <ImportForm/>
      : null
    }
    <button onClick={() => {}} className='export'>Export</button>
    <button onClick={toggleImportForm} className='toggle-import-form'>Toggle Import Form</button>
  </div>
);

SetlistBuilder.propTypes = {
  showImport: PropTypes.bool.isRequired,
  toggleImportForm: PropTypes.func.isRequired
}

export default SetlistBuilder;
