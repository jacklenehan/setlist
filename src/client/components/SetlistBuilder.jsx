import React from 'react';
import PropTypes from 'prop-types';

import ImportForm from '../containers/ImportForm';
import FilteredTrackList from '../containers/FilteredTrackList';

const SetlistBuilder = ({ lists, className, showImport, boundExportToText, toggleImportForm }) => (
  <div className={className}>
    {
      showImport
      ? <ImportForm />
      : null
    }
    {lists.map(list => <FilteredTrackList key={list.id} {...list}/>)}
    <button onClick={boundExportToText} className='export'>Export</button>
    <button onClick={toggleImportForm} className='toggle-import-form'>Toggle Import Form</button>
  </div>
)

SetlistBuilder.propTypes = {
  lists: PropTypes.array.isRequired,
  showImport: PropTypes.bool.isRequired,
  toggleImportForm: PropTypes.func.isRequired,
  boundExportToText: PropTypes.func.isRequired,
  className: PropTypes.string.isRequired
}

export default SetlistBuilder;
