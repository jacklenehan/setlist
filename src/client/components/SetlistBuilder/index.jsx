import React from 'react'
import PropTypes from 'prop-types'

import ImportForm from 'Client/components/ImportForm'
import FilteredTrackList from 'Client/components/FilteredTrackList'

const SetlistBuilder = ({ lists, className, showImport, boundExportToText, toggleImportForm }) => (
  <div className={className}>
    {showImport ? <ImportForm /> : null}
    {lists.map((list) => (
      <FilteredTrackList key={list.id} {...list} />
    ))}
    <button onClick={boundExportToText} className="export">
      Export
    </button>
    <button onClick={toggleImportForm} className="toggle-import-form">
      Toggle Import Form
    </button>
  </div>
)

SetlistBuilder.propTypes = {
  lists: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
    }),
  ).isRequired, // TODO: Shape right?
  showImport: PropTypes.bool.isRequired,
  toggleImportForm: PropTypes.func.isRequired,
  boundExportToText: PropTypes.func.isRequired,
  className: PropTypes.string.isRequired,
}

export default SetlistBuilder
