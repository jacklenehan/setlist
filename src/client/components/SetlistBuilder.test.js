import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import SetlistBuilder from './SetlistBuilder';
import ImportForm from '../containers/ImportForm';
import FilteredTrackList from '../containers/FilteredTrackList';

describe('SetlistBuilder', () => {
  let props;
  let wrapper;

  const setlistBuilder = () => {
    if(!wrapper) {
      wrapper = shallow(
        <SetlistBuilder {...props} />
      );
    }
    return wrapper;
  }

  beforeEach(() => {
    const toggleImportForm = jest.fn();
    const boundExportToText = jest.fn();

    props = {
      lists: [],
      showImport: false,
      className: 'foo',
      toggleImportForm,
      boundExportToText
    };

    wrapper = undefined;
  });

  describe('render', () => {
    it('renders without crashing', () => {
      setlistBuilder();
    });

    it('renders correctly', () => {
      const tree = renderer.create(<SetlistBuilder {...props} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('renders one FilteredTrackList for each list', () => {
      props.lists = [{maxDuration: 1, list: 'foo', id: 1}, {maxDuration: 1, list: 'bar', id: 2}];
      expect(setlistBuilder().find(FilteredTrackList).length).toBe(2);
    });

    it('renders the ImportForm when props.showImport is true', () => {
      props.showImport = true;
      expect(setlistBuilder().find(ImportForm).length).toBe(1);
    });
  });
});
