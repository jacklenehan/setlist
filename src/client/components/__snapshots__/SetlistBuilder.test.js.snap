// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SetlistBuilder render renders correctly 1`] = `
<div
  className="foo"
>
  <button
    className="export"
    onClick={[Function]}
  >
    Export
  </button>
  <button
    className="toggle-import-form"
    onClick={[Function]}
  >
    Toggle Import Form
  </button>
</div>
`;
